src/software/thunderscope/robot_diagnostics/handheld_device_manager.py

```diff
@@ -0,0 +1,338 @@
+        """This method sets all required fields in the power control proto to
+        the default/minimum default value
+        """
+        self.power_control = PowerControl()
+        self.power_control.geneva_slot = 3
+
+    def set_enabled(self, enabled) -> None:
+        """Enable or disable the HandheldDeviceManager from processing input events.
+
+        :param enabled: whether to enable (True) or disable (False) the
+                        HandheldDeviceManager from processing input events
+        """
+        with self.lock:
+            self.enabled = enabled
+
+    def reinitialize_handheld_device(self) -> None:
+        """Attempt to reinitialize a connection to a handheld device."""
+        self.__clear_handheld_device()
+        self.__initialize_handheld_device()
+
+    def __initialize_handheld_device(self) -> None:
+        """Attempt to initialize a connection to a handheld device.
+        The first handheld device that is recognized as valid will be used.
+        Valid handheld devices are any devices whose name has a matching DeviceConfig
+        in DiagnosticsConstants.HANDHELD_DEVICE_NAME_CONFIG_MAP.
+        """
+        device_config_map = DiagnosticsConstants.HANDHELD_DEVICE_NAME_CONFIG_MAP
+
+        for device in list_devices():
+            handheld_device = InputDevice(device)
+            if (
+                handheld_device is not None
+                and handheld_device.name in device_config_map
+            ):
+                with self.lock:
+                    self.handheld_device = handheld_device
+                    self.handheld_device_config = device_config_map[
+                        handheld_device.name
+                    ]
+
+                    self.handheld_device_connection_status_signal.emit(
+                        HandheldDeviceConnectionChangedEvent(
+                            HandheldDeviceConnectionStatus.CONNECTED,
+                            self.handheld_device.name,
+                        )
+                    )
+
+                    self.logger.debug("Successfully initialized handheld!")
+                    self.logger.debug(
+                        'Device name: "' + self.handheld_device.name + '"'
+                    )
+                    self.logger.debug("Device path: " + self.handheld_device.path)
+
+                return
+
+        self.__clear_handheld_device()
+        self.logger.debug("Failed to initialize handheld device, check USB connection")
+        self.logger.debug("Tried the following available devices:")
+        self.logger.debug(list(map(lambda d: InputDevice(d).name, list_devices())))
+
+    def __clear_handheld_device(self) -> None:
+        """Clears handheld device and config field by setting them to None,
+        and emits a disconnected notification signal.
+        """
+        with self.lock:
+            self.handheld_device = None
+            self.handheld_device_config = None
```

Comment:
why does the lock need to protect these fields?

Reply:
`reinitialize_handheld_device` is called on the Qt gui thread when the "detect handheld device" button is pressed. Without the lock, `HandheldDeviceManager`'s event loop thread can interleave with `reinitialize_handheld_device` (which mutates `self.handheld_device` and `self.handheld_device_config`)